### ISSUE TITLE
File transfer slow via tsclient when RDGateway is in use

### ISSUE THREAD

**Issue by aresliharris** [FreeRDP/FreeRDP#9983]
## Description of the bug
* A noticeable file transfer speed difference was captured from a project that I have been working on via Freerdp(xfreerdp). However, this is not happening while using Microsoft RDP client on Windows or Mac. 
* Specifically, when we transfer a 20mb file from a shared drive to a local hard disk on the remote server, which is the same server with RD Gateway, took over 3mins whereas 20s without using the gateway. (Robocopy was used for benchmarking in this test.)
* It has been verified that the specs of the server is not the bottle neck and we don't have additional security scanning going one during the test.
* The issue appears on both freerdp(xfreerdp) 2.7 and 3.2 
* We use Apache Guacamole, which uses freerdp 2.x on the latest version as of Mar 2014, and the same issue occurs during the test.

 
  ## The steps to reproduce the bug (a single 20MB test file transfer)
  * CLI command and configuration used
    * xfreerdp /v:<IP>:<port> /cert-ignore /dynamic-resolution /u:<username> /drive:<local_drive_path_for_tsclient_later> /g:<IP>:<port>
  * RDP server tested is Windows Server 2016 with RDGateway configured on the same server
  * Windows copy and paste tool were used to test (ctrl+C & ctrl+V)
  * Robocopy was also tested, but similar slow transfer performance was noticed
  * what did you expect to happen
    * Less than 20s completion time for the file transfer, which is similar to what was observed while using Microsoft RDP client with shared drive mounted via tsclient
  * what actually happened? 
    * It took around 3 mins to transfer the 20MB test file
  * freerdp version (e.g. xfreerdp --version) or package version or git commit.
    *  2.7.x and 3.2.x
  * Operating System, architecture, distribution e.g. linux, amd64, debian.
    * Workstation OS where the xfreerdp was run is on OS X 14.1.2

**Comment by akallabeth** [comment]
interesting find.

a few more questions:

1. which gateway protocol was used? (http, websocket or older rdg)
2. for the tests with windows/mac clients, did they use TCP or UDP?

**Comment by aresliharris** [comment]
> interesting find.
> 
> a few more questions:
> 
> 1. which gateway protocol was used? (http, websocket or older rpc)
> 2. for the tests with windows/mac clients, did they use TCP or UDP?

Hi @akallabeth , 

Thank you for commenting on this.

As you can see from the freerdp commend used, the slowness appeared when the connection was established based on the default protocol. From the Windows server side, I believe it's using websocket for the gateway protocol.

The tests with clients, they use TCP. Actually, both UDP and TCP were tested, but none of them made a difference on the file transfer issue.

Best,
Ares

**Comment by akallabeth** [comment]
@aresliharris ok.
can you try with explicit gateway config (disabled websocket and another test with rpc if supported by your config)?
would be interesting to narrow down if this is a generic issue or just for a specific gateway transport

**Comment by aresliharris** [comment]
> @aresliharris ok. can you try with explicit gateway config (disabled websocket and another test with rpc if supported by your config)? would be interesting to narrow down if this is a generic issue or just for a specific gateway transport

@akallabeth In the latest version freerdp 3.x, I tested all the transport types by using different flags, but none of them worked/performed normal speed file transfer from the shared tsclient drive to the RDP server connected. 

Would you be able to provide some guidance if you are talking about changing the gateway transport configs from the Windows server side? I tried to find ways to enforce the transport type from Windows server side, but I haven't been able to.

Thanks,
Ares

**Comment by akallabeth** [comment]
@aresliharris setting client side is enough.

1. `/gateway:type:http,no-websockets`
2. `/gateway:type:http`
3. `/gateway:type:rpc`

do you have any details on speed on each one of these?

it might be simple optimization of some packet sizes or it could be something more complicated.
don´t have time to do tests now though, we´re still in the bug fixing stage, so speed improvements usually come later

**Comment by aresliharris** [comment]
It performs normally when gateway is off as we discovered.

Hope this helps.

Thank you,
Ares

Ares Li
Team Lead, DevOps
***@***.***
+1 778-900-1373
www.cayenta.com
This message has been sent on behalf of a company that is part of the Harris Operating Group of Constellation Software Inc.
If you prefer not to be contacted by Harris Operating Group please notify us.

________________________________
From: akallabeth ***@***.***>
Sent: Wednesday, March 20, 2024 8:29 AM
To: FreeRDP/FreeRDP ***@***.***>
Cc: Ares Li ***@***.***>; Mention ***@***.***>
Subject: [EXTERNAL] Re: [FreeRDP/FreeRDP] File transfer slow via tsclient when RDGateway is in use (Issue #9983)


CAUTION: This email originated from outside of the organization. Do not click links or open attachments unless you recognize the sender and know the content is safe.


@aresliharris<https://urldefense.com/v3/__https://github.com/aresliharris__;!!HzYX23eHnp2Wbe4!RUO-jTmxSmq3IARvftEdKtYjMjx25N10VPbK8EHNKJB4V2i9vWslJfMZrA-T-jzQrhBsVN4cYqCf7f8WL-Kp$> oh, and most importantly, how does it perform for direct (non gateway) connections for you?

**Comment by RobLW** [comment]
Woo I thought this was an issue with the way our IT setup their RDP (something I have no control over unfortunately), I've been having to upload to web storage and download again on the remote system because file transfers has crazy slow. In fact it can even rent user input into a very delayed/laggy experience.
It seems to "hang" at the start like it's busy checking something before it'll actually start transferring.

When I tried `/gateway:type:http,no-websockets` it did indeed improve, not to levels you would hope for, but certainly better!
`/gateway:type:http` didn't offer any improvement. In fact in the end my remote screen went all black while I waited and connection timed out.
`/gateway:type:rpc` I am not allowed to connect like this, it seems to break some security rules and denies me access.
